Implemented ajax operations because I'm sick of the toggle callbacks.

Toflar opened the pull request:
Done :-) No more `toggleVisibilty` implementations :-D.
Checkout the docs for the usage :-) Thanks for the feedback.

/cc @qzminski @aschempp

qzminski commented:
Looks good to me :+1:

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,226 @@
+     * @param string         $action
+     * @param \DataContainer $dc
+     */
+    public function executePostActions($action, \DataContainer $dc)
```
public methods are usually ordered before private methods?

Toflar replied:
And so what? I like it better down there as it's called via hook :-)

aschempp commented:
Nice! The only things I would change are names. I would try to follow the core names, so
`states` => `options`
`states_callback` => `options_callback`
`ajax_callback` => `save_callback`

Also, there is no JS-fallback, right? Just asking because the core does have that.
Also missing is executing the `save_callback` on the value before saving, and a permission check if the user is allowed to toggle that field. Check out the core implementation for toggling (e.g. in `tl_article`).

Toflar commented:
> Nice! The only things I would change are names. I would try to follow the core names, so
> states => options
> states_callback => options_callback
> ajax_callback => save_callback

That's a really bad idea I find. First, it's not options, it's states. It does not have the same format as the `options` anyway because there's no label. Second, it makes it very hard to read in the DCA and causes possibly trouble when e.g. searching for `options_callback` you certainly don't want to receive those ones :-) 

> Also missing is executing the save_callback on the value before saving.

True, I'll add that :-) But here you already see the naming trouble: We would then have the haste save_callback executing the regular save_callback? I don't want to have this, really.

> Also, there is no JS-fallback, right? Just asking because the core does have that.

No, I always thought that was not necessary. One can still build it if they like.

> and a permission check if the user is allowed to toggle that field.

Yes! What about adding another callback there so if I just wanted to extend the permission check I don't have to copy the whole standard routine?

aschempp commented:
> First, it's not options, it's states. It does not have the same format as the options anyway because there's no label.

I disagree. Radio button values are also "states", so it's all the same. But the return value is fundamentally different (options = key/value).

> But here you already see the naming trouble: We would then have the haste save_callback executing the regular save_callback?

Is it actually a save_callback? Because according to the docs the `ajax_callback` it's supposed to return the next value?

I don't think we need extended permission check. The permissions are set in the user, they are on the field not the button.

Toflar commented:
* Renamed `states` to `options` and removed the callback (can be changed using the `onload_callback`).
* Removed `ajax_callback` and added support for `save_callback` which allows you to do additional things anyway.
* Added permission checks and an additional `check_permission_callback` for custom checks.
* Added support for versioning.
* Cleanup.

I think it's really cool now :) 

Regarding the core default implementation of the `checkPermissions()` calls: I think we did it a lot more intelligent ;-) The core just executes a normal ajax request on the very same page and the `button_callback` checks for presence of `tid` and if so it checks for permissions. However, this is totally useless (https://github.com/contao/core/blob/master/system/modules/core/dca/tl_article.php#L970) because the page is loaded (= everything is being constructed) and thus the `onload_callback` should be called anyway, right? So I don't know why it is called again explicitly here. Haste uses the `executePostActions` hook which gets called after the `DC_Table` object has been constructed (and thus the `onload_callback` is already called). So I don't need to consider this case as it's already done :)

aschempp commented on assets/haste.js:
```diff
@@ -0,0 +1,49 @@
+var Haste = {
```
```js
var Haste = Haste | {};

Haste.toggleAjaxOperation = function () …
```

Toflar replied:
Fixed in ea3cee95e0cb955df8ba2c499ada6e3baaab1c5f.

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+            return;
+        }
+
+        \Controller::loadDataContainer($dc->table);
```
This should be obsolete?

Toflar replied:
True, removed in fda5e968109786d3b843b4e9ae56e02376a0b051.

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+        // Check permissions
+        if (!$this->checkPermission($dc->table, $hasteAjaxOperationSettings)) {
+
+            \System::log(sprintf('Not enough permissions to toggle field %s::%s',
```
`sprintf` should be on the next line

Toflar replied:
Done in 359e4cc71499eb1f7c10223b47f391fe50453835.

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+        }
+
+        // Initialize versioning
+        $versions = new \Versions($dc->table, $id);
```
You must check if versioning is enabled for this table

Toflar replied:
No, the `Versions` class does this already :)

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+            ->execute($value, $id);
+
+        $versions->create();
+        \System::log(sprintf('A new version of record "%s.id=%s" has been created',
```
Is this not logged by the Version class?

Toflar replied:
Nope.

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+     */
+    private function addOnClickAttribute(&$operation)
+    {
+        $clickEventString = 'return Haste.toggleAjaxOperation(this, %s);';
```
I think you should add a `return false` to prevent the link action?

Toflar replied:
The `toggleAjaxOperation` returns false?

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+    {
+        $hasPermission = true;
+
+        if (!\BackendUser::getInstance()->hasAccess(
```
Not sure what happens if the field is not `exclude => true` ?

aschempp commented on library/Haste/Dca/AjaxOperations.php:
```diff
@@ -0,0 +1,303 @@
+     *
+     * @return bool
+     */
+    private function checkPermission($table, array $hasteAjaxOperationSettings)
```
Generally I think it would be better not to pass the config array but to pass the necessary values directly. This way the methods don't need to know about the config array setup. Applies to other methods, too

Toflar replied:
I prefer having the array, it's a lot better than having the values directly. Allows for a lot more flexibility as the method signature doesn't have to change. Anthony Ferrara once blogged about this (which is also the reason why the 3rd parameter is an array in the `password_hash` function but I cannot find the blog entry right now) :)

Also, here the whole array needs to be passed to the callback anyway.
